### DISCUSSION TITLE
Button / action to stop dGPU without disabling?

### DISCUSSION THREAD

**Discussion by m31-galaxy** [seerge/g-helper#1533] [Ideas]
Often I find myself wanting to stop processes keeping the dGPU awake, to reduce power draw, minimise temperatures and fan noise, etc.  but without switching to Eco mode. The reason for this is partially because switching to Eco mode and back takes a while, and partially because it's buggy - causing the dGPU to show up in the "Eject Removable Media" menu, screen freezing / going black, NVIDIA control panel not working, and worst of all an issue where the dGPU continues to draw power in Eco mode, a problem already known by e.g. XMG.

What are everyone else's thoughts on the matter?

**Comment by IceStormNG** [comment]
This will not work. You have to close all apps that locked the dGPU (can be seen in CMD via 'nvidia-smi') or somehow move them to the main GPU. Windows does that for some app after using a display that was connected to the NVGPU, like when using advanced Optimus or connecting directly to the dGPU via HDMI/DP.
Only way is to disable it which can result in those apps crashing.
If you have certain apps that use the dGPU but actually do not need them, you can use the Windows graphics settings to let them run on the iGPU instead.
Besides of that, this is the downside of hybrid graphics in windows. The experience is crap because MS wants OEMs to not do that in the first place.

**Reply by m31-galaxy** [reply]
I'm fine with apps that don't support switching from dGPU to iGPU crashing. The trouble is, Windows graphics settings just doesn't do the trick - some apps (like Discord) change executable paths on every update, and if I'm switching out of dGPU mode with Advanced Optimus then I can have 10-20 processes still lingering on the dGPU.

**Reply by IceStormNG** [reply]
That's a problem with Advanced Optimus and how windows handles it. Only solution is to either reboot or to not use Advanced optimus and use regular Optimus. Advanced Optimus has a lot of drawbacks because Windows moves apps to the GPU that run the display of their app window.

**Reply by m31-galaxy** [reply]
Working what what we've got, instead of just "don't use this" would it not be better to include a button or action to do this? I believe it's already a feature, it's just hidden for devices that support switching to Eco mode, unless I'm misremembering.

**Reply by IceStormNG** [reply]
While I generally agree with you, the "kill all apps" has some issues. The reason is, that windows moves some core processes to the dGPU, killing them can corrupt the icon or search cache or breaking some other features.
Switching to Eco mode disables the GPU, which forces Windows to move the apps to the other GPU, which works fine in most cases.

**Reply by m31-galaxy** [reply]
The problem with Eco mode is that, ironically, it can increase power consumption. After doing a bit of digging online, it seems this is a [well-established phenomenon](https://www.reddit.com/r/XMG_gg/comments/m5fiqb/bug_disabling_dgpu_in_device_manager_increases/), and the XMG laptop company explicitly advises against disabling the dGPU through a method equivalent to disabling it in Task Manager.

I've chucked together a hacky fix of running a PowerShell script which disables and re-enables the dGPU in Device Manager - since you seem to know your way around this stuff way better than I do, does that raise any immediate concerns to you?

**Reply by IceStormNG** [reply]
Eco mode on ASUS laptops is not the same as what XMGs do. Eco mode on ASUS Laptops electrically disconnects the dGPU. It doesn't even show up anymore in device manager anymore. it does not disable the dGPU in Device manger. What you find disabled there is the platform hub controller. A device that is very rarely found in laptops.

The PS1 script can do the trick. It might screw up the driver at some point though. Nothing a reboot cannot fix.

There isn't a really simple solution because the problem roots deeply in Windows and the graphics stack. Advanced Optimus is a super hacky solution. What it does is, it disconnects the internal display from the iGPU and connects it to the dGPU (while leaving a disabled connection to the iGPU there). This moves all apps to the dGPU framebuffer, sometimes it does not, which causes massive stutters in those apps. Optimus only works well in one direction. dGPU -> iGPU. Not the other way round.
When you go back to optimius, it leaves most apps on the dGPU because the GPU is still there and how optimus works. And moves the others back to the iGPU.
The Windows preference for which GPU to use only works well for DirectX apps or standard Win32 Apps. Anything that uses Vulkan or OpenGL might or might not ignore it. Apps that use Nvidia specific features, like NVEnc or CUDA will ignore that, too.

**Reply by m31-galaxy** [reply]
I see, thank you for the detailed explanation. Laptops can be a real pain sometimes...

After doing some more testing, it does seem like my power draw is higher in Eco mode than in Standard mode. That doesn't seem right if the dGPU is being disconnected at a hardware level - is there some other underlying issue or am I maybe getting dodgy readings or missing some other key factor? A -35W power draw is not acceptable for me when using battery saver, Eco mode, 60Hz refresh on minimum brightness and a 10W CPU power limit.
